ee/modules/502-delivery/hooks/werf_sources/hook.go

```diff
@@ -0,0 +1,403 @@
+		}
+
+		argoRepos = append(argoRepos, argocdHelmOCIRepository{
+			Name:     ws.Name,
+			Username: creds.username,
+			Password: creds.password,
+			Project:  ws.ArgocdRepo.Project,
+			URL:      ws.Repo,
+		})
+	}
+	return argoRepos, nil
+}
+
+func extractCredentials(credentialsBySecretName map[string]dockerFileConfig, pullSecretName, registry string) (registryCredentials, error) {
+	creds := registryCredentials{}
+	if pullSecretName == "" {
+		// No credentials is OK for public registries, there can be unspecified pullsecret
+		return creds, nil
+	}
+
+	config, ok := credentialsBySecretName[pullSecretName]
+	if !ok {
+		return creds, fmt.Errorf("unknown pull secret %q", pullSecretName)
+	}
+
+	cfg, ok := config.Auths[registry]
+	if !ok {
+		return creds, fmt.Errorf("no credentials")
+	}
+
+	if cfg.Auth != "" {
+		auth, err := base64.StdEncoding.DecodeString(cfg.Auth)
+		if err != nil {
+			return creds, fmt.Errorf(`cannot decode base64 "auth" field`)
+		}
+		parts := strings.Split(string(auth), ":")
+		if len(parts) != 2 {
+			return creds, fmt.Errorf(`unexpected format of "auth" field, expected "username:password"`)
+		}
+		creds.username, creds.password = parts[0], parts[1]
+		return creds, nil
+	}
+
+	creds.username, creds.password = cfg.Username, cfg.Password
+	return creds, nil
+}
+
+// cr.example.com/path/to/image -> cr.example.com
+func firstSegment(s string) string {
+	for i, c := range s {
+		if c == '/' {
+			return s[:i]
+		}
+	}
+	return s
+}
+
+func filterWerfSource(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var (
+		ws  werfSource
+		err error
+		ok  bool
+	)
+
+	ws.Name = obj.GetName()
+
+	ws.Repo, ok, err = unstructured.NestedString(obj.Object, "spec", "imageRepo")
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		return nil, fmt.Errorf("spec.imageRepo field expected")
+	}
+
+	ws.APIURL, ok, err = unstructured.NestedString(obj.Object, "spec", "apiURL")
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		ws.APIURL = "https://" + firstSegment(ws.Repo)
+	}
+
+	ws.PullSecretName, _, err = unstructured.NestedString(obj.Object, "spec", "pullSecretName")
+	if err != nil {
+		return nil, err
+	}
+
+	// By default, Argo CD is desired, but the OCI repo can be disabled to use purely Image
+	// Updater functionality along with another repository type.
+	repoEnabled, ok, err := unstructured.NestedBool(obj.Object, "spec", "argocdRepoEnabled")
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		repoEnabled = true
+	}
+
+	// By default, Argo CD repo belongs to the "default" project.
+	arepo, ok, err := unstructured.NestedStringMap(obj.Object, "spec", "argocdRepo")
+	if err != nil {
+		return nil, err
+	}
+	project := "default"
+	if repoEnabled && ok {
+		specifiedProject, projectSpecified := arepo["project"]
+		if projectSpecified && specifiedProject != "" {
+			project = specifiedProject
+		}
+	}
+	if repoEnabled {
+		ws.ArgocdRepo = &argocdRepoConfig{
+			Project: project,
+		}
+	}
+
```

Comment:
Errors from FilterFn are handled by the separate go routines and they visible in logs only.
The better place to validate WerfSource and apply defaults is the `parseWerfSources` method. It is called during the hook run, what makes it possible to enter an error-loop state or update /status for the WerfSource.

Reply:
I moved all the verification to the hook. Thanks!